externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/FlinkSQLEngine.scala

```diff
@@ -103,4 +112,37 @@ object FlinkSQLEngine extends Logging {
       addShutdownHook(() => engine.stop(), FLINK_ENGINE_SHUTDOWN_PRIORITY + 1)
     }
   }
+
+  private def discoverDependencies(
+      jars: Seq[URL],
+      libraries: Seq[URL]): List[URL] = {
+    try {
+      var dependencies: List[URL] = List()
+      // find jar files
+      jars.foreach { url =>
+        JarUtils.checkJarFile(url)
+        dependencies = url :: dependencies
+      }
+      // find jar files in library directories
+      libraries.foreach { libUrl =>
+        val dir: File = new File(libUrl.toURI)
+        if (!dir.isDirectory) throw new SqlClientException("Directory expected: " + dir)
+        else if (!dir.canRead) throw new SqlClientException("Directory cannot be read: " + dir)
+        val files: Array[File] = dir.listFiles
+        if (files == null) throw new SqlClientException("Directory cannot be read: " + dir)
+        files.foreach { f =>
+          // only consider jars
+          if (f.isFile && f.getAbsolutePath.toLowerCase.endsWith(".jar")) {
+            val url: URL = f.toURI.toURL
+            JarUtils.checkJarFile(url)
+            dependencies = url :: dependencies
```

Comment:
Do we reverse the order of the files here by concatenating the elements at the head? I think the order matters, especially when there are potential class conflicts (maybe we should sort the files alphabetically as well).

Reply:
@link3280, I don't thank the order matters because the potential class conflicts couldn't be avoided by the order.